<a href="https://colab.research.google.com/github/Jialan8020/Assign8/blob/master/question4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Fit a multilayer neural network model on the SHHS data to predict whether rdi4p > 7 using the variables from the previous HW assigment. Compare the test set accuracy with the penalized model from your previous HW.

In [0]:
import pandas as pd
import tensorflow as tf
import numpy as np
from sklearn.model_selection import cross_val_score
from tensorflow import keras
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
from sklearn.metrics import accuracy_score

In [62]:
dat = pd.read_csv("shhs.txt",sep="\t")
dat = dat.dropna()
dat.head(4)

pptid  waist  COPD15  ASTHMA15  ...  age_s1  smokstat_s1   WASO     bmi_s1
3       4   85.0     0.0       0.0  ...      48          0.0   43.5  20.185185
4       5   76.0     0.0       0.0  ...      66          2.0  100.5  23.309053
8       9   99.0     0.0       0.0  ...      69          2.0   52.5  25.817447
14     15   80.0     0.0       0.0  ...      58          2.0  132.5  23.937672

[4 rows x 30 columns]

In [63]:
# Extract the rdi4p column
rdi = pd.DataFrame(dat["rdi4p"])
rdi

rdi4p
3      0.797342
4      2.756757
8     11.948791
14    16.254980
15     3.102041
...         ...
5795   2.108346
5798   0.469667
5799  34.415781
5801   2.163462
5802   1.038319

[2897 rows x 1 columns]

In [64]:
rdi[rdi < 7] = 0
rdi[rdi > 7] = 1
rdi = rdi.astype(int)
rdi


rdi4p
3         0
4         0
8         1
14        1
15        0
...     ...
5795      0
5798      0
5799      1
5801      0
5802      0

[2897 rows x 1 columns]

In [65]:
X = dat.drop(["rdi4p","pptid"], axis=1)
X

waist  COPD15  ASTHMA15  slp_lat  ...  age_s1  smokstat_s1   WASO     bmi_s1
3      85.0     0.0       0.0     14.0  ...      48          0.0   43.5  20.185185
4      76.0     0.0       0.0      6.5  ...      66          2.0  100.5  23.309053
8      99.0     0.0       0.0     20.0  ...      69          2.0   52.5  25.817447
14     80.0     0.0       0.0     66.5  ...      58          2.0  132.5  23.937672
15    100.0     0.0       0.0     69.5  ...      60          0.0   84.0  29.475309
...     ...     ...       ...      ...  ...     ...          ...    ...        ...
5795  117.0     0.0       0.0     29.5  ...      71          0.0   55.5  35.790598
5798  103.0     0.0       0.0     10.5  ...      54          2.0   13.0  32.414213
5799   99.0     0.0       0.0     11.5  ...      66          0.0   98.5  25.693134
5801   91.0     0.0       0.0      4.0  ...      55          0.0   30.0  24.228571
5802   93.0     0.0       0.0      8.0  ...      53          0.0   76.0  25.374483

[2897 rows x 28 columns]

In [66]:
# Separate training and test data
X_train, X_test, y_train, y_test = train_test_split(X, rdi, test_size=0.2)
scaler = preprocessing.StandardScaler().fit(X_train)
scaler.transform(X_train)
scaler.transform(X_test)

array([[-1.21685054, -0.09619785, -0.19617508, ..., -1.00264747,
         2.22488609, -1.28191779],
       [-0.85238048, -0.09619785, -0.19617508, ..., -1.00264747,
        -0.54685562,  0.05643812],
       [-0.05783575, -0.09619785, -0.19617508, ...,  0.0466453 ,
        -0.22938741, -0.47774143],
       ...,
       [-1.72710863, -0.09619785, -0.19617508, ...,  1.09593806,
         2.5423543 , -0.89344508],
       [-0.29109659, -0.09619785, -0.19617508, ...,  1.09593806,
         2.3836202 ,  0.21692731],
       [-0.3421224 , -0.09619785, -0.19617508, ..., -1.00264747,
         0.61312439,  0.01360812]])

In [0]:
model = tf.keras.Sequential([
  tf.keras.layers.Dense(64, activation='relu',input_dim=28),
  tf.keras.layers.Dense(10, activation='relu'),
  tf.keras.layers.Dense(1,activation='sigmoid')
])
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [72]:
model.fit(X_train, y_train, epochs=5)

Epoch 1/5
73/73 [==============================] - 0s 1ms/step - loss: 0.7481 - accuracy: 0.6280
Epoch 2/5
73/73 [==============================] - 0s 1ms/step - loss: 0.9455 - accuracy: 0.6271
Epoch 3/5
73/73 [==============================] - 0s 1ms/step - loss: 0.7136 - accuracy: 0.6491
Epoch 4/5
73/73 [==============================] - 0s 1ms/step - loss: 0.6716 - accuracy: 0.6599
Epoch 5/5
73/73 [==============================] - 0s 1ms/step - loss: 0.6986 - accuracy: 0.6543


In [73]:
model.evaluate(X_test, y_test)

19/19 [==============================] - 0s 1ms/step - loss: 0.7489 - accuracy: 0.6086


[0.7489436864852905, 0.6086207032203674]

**Logistic Regression**

In [0]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_squared_error


L1 (Lasso regression

In [76]:
model2 = LogisticRegression(solver='liblinear',penalty="l1")
model2.fit(X_train, y_train)
predict_2 = model2.predict(X_test)
accuracy_score(y_test, predict_2)
# mse2 = mean_squared_error(y_test, predict_2)
# mse2

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.7034482758620689

L2 penalty (Ridge regression)

In [77]:
model3 = LogisticRegression(solver='liblinear',penalty="l2")
model3.fit(X_train, y_train)
predict_3 = model3.predict(X_test)
accuracy_score(y_test, predict_3)
# mse3 = mean_squared_error(y_test, predict_3)
# mse3

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.6982758620689655

Elastic net regression

In [78]:
model4 = LogisticRegression(solver='saga',penalty="elasticnet",l1_ratio=0.5)
model4.fit(X_train, y_train)
predict_4 = model4.predict(X_test)
accuracy_score(y_test, predict_4)
# mse4 = mean_squared_error(y_test, predict_4)
# mse4

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


0.6793103448275862

No penalty

In [79]:
model5 = LogisticRegression(penalty="none",solver="newton-cg")
model5.fit(X_train, y_train)
predict_5 = model5.predict(X_test)
accuracy_score(y_test, predict_5)
# mse5 = mean_squared_error(y_test, predict_5)
# mse5

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.6913793103448276

**Comparison with HW 7**

The best cross validated accuracy of Lasso regression is 0.703.

The best cross validated accuracy of Ridge regression is 0.698.

The best cross validated accuracy of elastic net is 0.679.

The best cross validated accuracy of no penalty is 0.691.

By using neural network algorithm, we achieved an accuracy of 0.609, which is significantly smaller than the logistic regression model.

The Lasso Regression yielded the best results.
